src/lit-element.ts

```diff
@@ -58,59 +60,77 @@ export class LitElement extends UpdatingElement {
    */
-  static styles?: CSSResult|CSSResultArray;
+  static styles?: CSSResultOrNative|CSSResultArray;
 
   private static _styles: CSSResult[]|undefined;
+  private static _nativeStyles: CSSStyleSheet[]|undefined;
 
   /**
    * Return the array of styles to apply to the element.
    * Override this method to integrate into a style management system.
    *
    * @nocollapse
    */
-  static getStyles(): CSSResult|CSSResultArray|undefined {
+  static getStyles(): CSSResultOrNative|CSSResultArray|undefined {
     return this.styles;
   }
 
   /** @nocollapse */
   private static _getUniqueStyles() {
     // Only gather styles once per class
-    if (this.hasOwnProperty(JSCompiler_renameProperty('_styles', this))) {
+    const property = supportsAdoptingShadowStyleSheets ?
+        JSCompiler_renameProperty('_nativeStyles', this) :
+        JSCompiler_renameProperty('_styles', this);
+    if (this.hasOwnProperty(property)) {
       return;
     }
     // Take care not to call `this.getStyles()` multiple times since this
     // generates new CSSResults each time.
-    // TODO(sorvell): Since we do not cache CSSResults by input, any
-    // shared styles will generate new stylesheet objects, which is wasteful.
-    // This should be addressed when a browser ships constructable
-    // stylesheets.
     const userStyles = this.getStyles();
-    if (userStyles === undefined) {
-      this._styles = [];
-    } else if (Array.isArray(userStyles)) {
+    const work = [];
+
+    if (Array.isArray(userStyles)) {
       // De-duplicate styles preserving the _last_ instance in the set.
       // This is a performance optimization to avoid duplicated styles that can
       // occur especially when composing via subclassing.
       // The last item is kept to try to preserve the cascade order with the
       // assumption that it's most important that last added styles override
       // previous styles.
       const addStyles =
-          (styles: CSSResultArray, set: Set<CSSResult>): Set<CSSResult> =>
+          (styles: CSSResultArray, set: Set<CSSResultOrNative>): Set<CSSResultOrNative> =>
               styles.reduceRight(
-                  (set: Set<CSSResult>, s) =>
+                  (set: Set<CSSResultOrNative>, s) =>
                       // Note: On IE set.add() does not return the set
                   Array.isArray(s) ? addStyles(s, set) : (set.add(s), set),
                   set);
       // Array.from does not work on Set in IE, otherwise return
       // Array.from(addStyles(userStyles, new Set<CSSResult>())).reverse()
-      const set = addStyles(userStyles, new Set<CSSResult>());
-      const styles: CSSResult[] = [];
-      set.forEach((v) => styles.unshift(v));
-      this._styles = styles;
+      const set = addStyles(userStyles, new Set<CSSResultOrNative>());
+      set.forEach((v) => work.unshift(v));
+    } else if (userStyles !== undefined) {
+      work.push(userStyles);
+    }
+
+    if (supportsAdoptingShadowStyleSheets) {
```

Comment:
Can we do one loop with this check inside? I think it'll be more compact.

Reply:
We're creating two different arrays. I could change the `_styles` and `_nativeStyles` to be a type union but I think it'll be more confusing when we apply them in `adoptStyles`.